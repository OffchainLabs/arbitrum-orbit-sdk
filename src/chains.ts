import { defineChain } from 'viem';
import {
  mainnet,
  holesky,
  arbitrum as arbitrumOne,
  arbitrumNova,
  sepolia,
  arbitrumSepolia,
} from 'viem/chains';

const nitroTestnodeL1 = defineChain({
  id: 1_337,
  network: 'Nitro Testnode L1',
  name: 'nitro-testnode-l1',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8545'],
    },
    public: {
      http: ['http://127.0.0.1:8545'],
    },
  },
  testnet: true,
});

const nitroTestnodeL2 = defineChain({
  id: 412_346,
  network: 'Nitro Testnode L2',
  name: 'nitro-testnode-l2',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8547'],
    },
    public: {
      http: ['http://127.0.0.1:8547'],
    },
  },
  testnet: true,
});

// the L3 deployed as part of the nitro-testnode framework uses a custom fee token
const nitroTestnodeL3 = defineChain({
  id: 333333,
  network: 'Nitro Testnode L3',
  name: 'nitro-testnode-l3',
  nativeCurrency: { name: 'AppTestToken', symbol: 'APP', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:3347'],
    },
    public: {
      http: ['http://127.0.0.1:3347'],
    },
  },
  testnet: true,
});

export const chains = [
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
] as const;

export {
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
};
