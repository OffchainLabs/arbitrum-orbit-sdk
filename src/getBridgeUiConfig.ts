import { createPublicClient, PublicClient, Chain, Transport, http } from 'viem';
import { ChainConfig } from './types/ChainConfig';
import { BridgeUiConfig } from './types/bridgeUiConfig';
import { createRollupPrepareTransaction } from './createRollupPrepareTransaction';
import { createRollupPrepareTransactionReceipt } from './createRollupPrepareTransactionReceipt';
import { createTokenBridgeFetchTokenBridgeContracts } from './createTokenBridgeFetchTokenBridgeContracts';

/**
 * Parameters for retrieving bridge UI configuration
 */
export type GetBridgeUiConfigParams<TChain extends Chain> = {
  parentChain: TChain;
  deploymentTxHash: `0x${string}`;
  chainName?: string;
  rpcUrl?: string;
  explorerUrl?: string;
  parentChainPublicClient?: PublicClient<Transport, TChain>;
};

/**
 * Retrieves the bridge UI configuration for an Orbit chain
 *
 * This function fetches all necessary contract addresses and chain information
 * required to set up a bridge UI for an Orbit chain. It uses the deployment
 * transaction to extract core contract addresses and token bridge information.
 *
 * @param params - Configuration parameters for retrieving bridge UI information
 * @returns Bridge UI configuration information
 *
 * @example
 * ```typescript
 * const bridgeConfig = await getBridgeUiConfig({
 *   parentChain: arbitrumSepolia,
 *   deploymentTxHash: '0x...',
 *   chainName: 'My Orbit Chain',
 *   rpcUrl: 'https://my-rpc-url',
 *   explorerUrl: 'https://my-explorer'
 * });
 * ```
 */
export async function getBridgeUiConfig<TChain extends Chain>(
  params: GetBridgeUiConfigParams<TChain>,
): Promise<BridgeUiConfig> {
  const {
    parentChain,
    deploymentTxHash,
    chainName = 'Orbit Chain',
    rpcUrl = 'http://localhost:8449',
    explorerUrl = 'http://localhost',
    parentChainPublicClient: providedClient,
  } = params;

  // Create a new public client if not provided
  const parentChainPublicClient: PublicClient<Transport, TChain> =
    providedClient ||
    createPublicClient({
      chain: parentChain,
      transport: http(),
    });

  const txData = await parentChainPublicClient.getTransaction({ hash: deploymentTxHash });
  // Get the deployment transaction
  const tx = createRollupPrepareTransaction(txData);

  // Extract chain configuration from transaction inputs
  const config = tx.getInputs()[0].config;
  const chainConfig: ChainConfig = JSON.parse(config.chainConfig);

  // Get the transaction receipt
  const txReceiptData = await parentChainPublicClient.getTransactionReceipt({
    hash: deploymentTxHash,
  });
  const txReceipt = createRollupPrepareTransactionReceipt(txReceiptData);

  // Get core contract addresses
  const coreContracts = txReceipt.getCoreContracts();

  // Get token bridge contract addresses
  const tokenBridgeContracts = await createTokenBridgeFetchTokenBridgeContracts({
    inbox: coreContracts.inbox,
    parentChainPublicClient,
  });

  // Build and return the configuration object
  return {
    chainInfo: {
      chainName,
      chainId: chainConfig.chainId,
      parentChainId: parentChain.id,
      rpcUrl,
      explorerUrl,
      nativeToken: coreContracts.nativeToken,
    },
    coreContracts: {
      rollup: coreContracts.rollup,
      inbox: coreContracts.inbox,
      outbox: coreContracts.outbox,
      sequencerInbox: coreContracts.sequencerInbox,
      bridge: coreContracts.bridge,
      nativeToken: coreContracts.nativeToken,
    },
    tokenBridgeContracts: {
      l2Contracts: {
        customGateway: tokenBridgeContracts.parentChainContracts.customGateway,
        multicall: tokenBridgeContracts.parentChainContracts.multicall,
        proxyAdmin: tokenBridgeContracts.orbitChainContracts.proxyAdmin,
        router: tokenBridgeContracts.parentChainContracts.router,
        standardGateway: tokenBridgeContracts.parentChainContracts.standardGateway,
        weth: tokenBridgeContracts.parentChainContracts.weth,
        wethGateway: tokenBridgeContracts.parentChainContracts.wethGateway,
      },
      l3Contracts: {
        customGateway: tokenBridgeContracts.orbitChainContracts.customGateway,
        multicall: tokenBridgeContracts.orbitChainContracts.multicall,
        proxyAdmin: tokenBridgeContracts.orbitChainContracts.proxyAdmin,
        router: tokenBridgeContracts.orbitChainContracts.router,
        standardGateway: tokenBridgeContracts.orbitChainContracts.standardGateway,
        weth: tokenBridgeContracts.orbitChainContracts.weth,
        wethGateway: tokenBridgeContracts.orbitChainContracts.wethGateway,
      },
    },
  };
}
