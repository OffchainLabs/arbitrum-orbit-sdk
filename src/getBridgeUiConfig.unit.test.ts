import { it, expect } from 'vitest';
import { createPublicClient, http } from 'viem';
import { sepolia } from 'viem/chains';

import { getBridgeUiConfig } from './getBridgeUiConfig';

const sepoliaClient = createPublicClient({
  chain: sepolia,
  transport: http('https://sepolia.gateway.tenderly.co'),
});

// https://sepolia.etherscan.io/tx/0xc172278b80251a7368409b5358fbf7e6250c1afc9a262e38ca15e1ba2bb9a7a8
it('successfully generates bridge UI config from deployment transaction', async () => {
  const bridgeUiConfig = await getBridgeUiConfig({
    params: {
      parentChain: sepolia,
      deploymentTxHash: '0xc172278b80251a7368409b5358fbf7e6250c1afc9a262e38ca15e1ba2bb9a7a8',
    },
    parentChainPublicClient: sepoliaClient,
  });

  expect(bridgeUiConfig).toMatchSnapshot();
});
