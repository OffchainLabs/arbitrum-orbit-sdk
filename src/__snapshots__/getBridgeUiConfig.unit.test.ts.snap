// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`successfully generates bridge UI config from deployment transaction 1`] = `
{
  "chainInfo": {
    "chainId": 421614,
    "chainName": "My Orbit Chain",
    "explorerUrl": "http://localhost:4000",
    "nativeToken": "0x0000000000000000000000000000000000000000",
    "parentChainId": 11155111,
    "rpcUrl": "http://localhost:8449",
  },
  "coreContracts": {
    "bridge": "0x683aa7Bf1E4346d8896720c1E51901DCE83cE96b",
    "inbox": "0x407b7C32eD620b37F60BBCd9a746B5a575c9FfaB",
    "nativeToken": "0x0000000000000000000000000000000000000000",
    "outbox": "0x3b7B4020AA8d20C565a7B1d2475db34F2b00D396",
    "rollup": "0x8D31CB7e6A630F3732614d382c7EB9F3833DBb86",
    "sequencerInbox": "0xc4b2140fD7B513148E9C1C0c53cA5C7AEefdc87A",
  },
  "tokenBridgeContracts": {
    "l2Contracts": {
      "customGateway": "0xB59dD42cC815FdA12B8b40A410164D49Fd2E96Fc",
      "multicall": "0x73465577E9FD7Cd585E4270F23A9eBa99B92b6eD",
      "proxyAdmin": "0x7618B3d4d8C71132BfCd33fDD98424a060F37470",
      "router": "0xFb4ea97964A04D0Dd367Bd69ffF03e66dfDB55b8",
      "standardGateway": "0x6b6f76e5D2A2C8Ee2cc2964ccaAFf570835B003a",
      "weth": "0x7b79995e5f793A07Bc00c21412e50Ecae098E7f9",
      "wethGateway": "0x350Ac26Db427d820cEFaF7Cf50Ae92046217bF06",
    },
    "l3Contracts": {
      "customGateway": "0x8399A1F530Ea292Fb3669cDa56b11bD87938b5d7",
      "multicall": "0xb4FBb198D9308126d9A3cdbff9d986836d53A894",
      "proxyAdmin": "0x7618B3d4d8C71132BfCd33fDD98424a060F37470",
      "router": "0x301987Bb8fAA1bB19411A7Bb3F74D387A7916b8c",
      "standardGateway": "0x0eA2d857a49CfF06c45271a970396D2B1dBd1382",
      "weth": "0x4787961Dd15d9B278Cb8a8E9c3e3ef079f1849cB",
      "wethGateway": "0xBabc9E20db02c8E9E0556826b0A1eF6c24F4C049",
    },
  },
}
`;
